File: 백준/Gold/6951. Packet Routing/Packet Routing.py
# https://www.acmicpc.net/problem/6951
import sys
input = sys.stdin.readline
INF = float('inf')

n, w, p = map(int, input().split())
graph = [[INF]*(n+1) for _ in range(n+1)]

for i in range(n+1):
    graph[i][i] = 0
    
for _ in range(w):
    u, v, w = map(int, input().split())
    graph[u][v] = w
    graph[v][u] = w

for k in range(1, n+1):
    for i in range(1, n+1):
        for j in range(1, n+1):
            if graph[i][k] != INF and graph[k][j] != INF:
                graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j])

print('\n'.join(map(lambda ipt: str(graph[ipt[0]][ipt[1]]), [list(map(int, input().split())) for _ in range(p)])))

File: 백준/Gold/6951. Packet Routing/Packet Routing.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.stream.IntStream;


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int INF = Integer.MAX_VALUE;

        int[] nwp = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int n = nwp[0];
        int w = nwp[1];
        int p = nwp[2];

        int[][] graph = new int[n+1][n+1];
        for (int[] line: graph)
            Arrays.fill(line, INF);
        for (int i = 0; i < n+1; i++)
            graph[i][i] = 0;

        while (w-- > 0) {
            int[] abc = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();

            int a = abc[0];
            int b = abc[1];
            int c = abc[2];

            graph[a][b] = c;
            graph[b][a] = c;
        }

        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    if (graph[i][k] != INF && graph[k][j] != INF)
                        graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
                }
            }
        }

        StringBuilder ans = new StringBuilder();
        while (p-- > 0) {
            int[] ab = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();

            int a = ab[0];
            int b = ab[1];

            ans.append(graph[a][b]).append('\n');
        }

        System.out.println(ans);
    }
}